# Проект по игре "Секреты Темнолесья" (рынок видеоигр)

- Автор: Козлова Екатерина
- Дата: 18.01.2026 - 20.01.2026

### Цели и задачи проекта

**Цель проекта** - исследовать развитие игровой индустрии в 2000-2013 гг. (платформы, продажи по жанрам, предпочтения и т.д.)

**Задачи проекта:**
1. Загрузка данных и первичное знакомство с ними
2. Предобработка данных, включая проверку корректности данных, их типов, наличия или отсуствия пропусков
3. Фильтрация данных по временному периоду
4. Категоризация данных

### Описание данных


Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
* Name - название игры.

* Platform - название платформы.

* Year of Release - год выпуска игры.

* Genre - жанр игры.

* NA sales - продажи в Северной Америке (в миллионах проданных копий).

* EU sales - продажи в Европе (в миллионах проданных копий).

* JP sales - продажи в Японии (в миллионах проданных копий).

* Other sales - продажи в других странах (в миллионах проданных копий).

* Critic Score - оценка критиков (от 0 до 100).

* User Score - оценка пользователей (от 0 до 10).

* Rating - рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта


* [Загрузка данных и знакомство с ними](#1bullet) 
* [Проверка ошибок в данных и их предобработка](#2bullet)
* [Фильтрация данных](#3bullet)
* [Категоризация данных](#4bullet)

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
# Пустые ячейки после каждого задания — примерное пространство для работы.
# Вы можете свободно добавлять или удалять ячейки по своему усмотрению в зависимости от логики и объёма работы.

In [2]:
# Используйте ячейки типа Code для вашего кода

In [3]:
# При необходимости добавьте новые ячейки для кода

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [4]:
import pandas as pd

In [5]:
df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

In [6]:
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

<font color='#777778'>Используйте ячейки типа Markdown для промежуточных выводов и расширенных комментариев к действиям с данными. </font>

**Выводы на основании полученных данных:**

Всего в датафрейме 16956 строк и 11 столбцов.

Представленные жанные в целом соответствуют ожидаемой структуре (название игры, платформы, продажи по регионам, оценки и рейтинги).

Больше всего пропусков в столбцах Critic Score (8242), User Score (10152), Rating (10085). В столбцах Name и Genre по 2 пропуска. И незначительное количество в столбце Year of Release (275).

Столбцы Name, Plaform, Genre и Rating имеют верный тип данных - object (однако их необходимо будет проверить на наличие дубликатов и верно ли использованы регистры). Столбцы NA sales, Other sales, Critic Score также имеют корректный тип, а именно float64. 

В некоторых столбцах имеются недочеты:
- столбец Year of Release - поскольку столбец содержит год выпуска игры, правильнее, чтобы он был преобразован либо в int64, либо в datetime64;
- столбцы EU sales и JP sales - их правильнее привести к типу float64 (возможно, они содержат строки, из-за чего тип данных был некорректно применен)
- столбец User Score - оценку пользователей корректнее привести к типу float64 (возможно, он также содержит строки)

В целом все столбцы соответствуют ожиданию. Однако, для удобства работы с Python, столбцы корректнее перевести в cnak case. Сокращения NA, EU, JP - могут быть не всем понятны, поэтому можно добавить описание данных сокращений. 

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [8]:
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [9]:
df.columns = df.columns.str.lower().str.replace(' ', '_')

In [10]:
display(df)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...
16951,Samurai Warriors: Sanada Maru,PS3,2016.0,Action,0.00,0.0,0.01,0.00,NaN,NaN,NaN
16952,LMA Manager 2007,X360,2006.0,Sports,0.00,0.01,0.0,0.00,NaN,NaN,NaN
16953,Haitaka no Psychedelica,PSV,2016.0,Adventure,0.00,0.0,0.01,0.00,NaN,NaN,NaN
16954,Spirits & Spells,GBA,2003.0,Platform,0.01,0.0,0.0,0.00,NaN,NaN,NaN


После приведения столбцов к snak case достигнута единообразность именования и все столбцы соответствуют общепринятому стандарту Python. Поскольку убраны пробелы, то к столбцам теперь проще обращаться, нет необходимости брать их в кавычки.

In [11]:
tmp = df.copy() 
len(tmp)

16956

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16956 non-null  object 
 6   jp_sales         16956 non-null  object 
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


Можно заметить, что в каждом из этих столбцов присутствует nan, соответственно Python не может распознать тип и приводит его к универсальному типу object.

In [13]:
for column in ['year_of_release', 'eu_sales', 'jp_sales', 'user_score']:
    df[column] = pd.to_numeric(df[column], errors='coerce')

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB


В начале проектной работы было выявлено, что ряд столбцов (а именно year_of_release, eu_sales, jp_sales, user_score) имеют неверный тип данных. Благодаря коду выше в данных столбцах внесены необходимые коррективы. 

### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


*Посчитаем абсолютное количество пропусков в каждом столбце*

In [15]:
df.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

*Посчитаем относительное количество пропусков в каждом столбце*

In [16]:
df.isna().sum() / len(df) * 100

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64

Самыми надежными столбцами для анализа являются столбцы platform, na_sales, other_sales, name, genre, т.к. в них 0-2 пропуска. Незначительное количество пропусков хранится в столбцах eu_sales (6) и jp_sales (4). Данные пропуски можно либо удалить, либо заменить недостающими значениями после обработки. Относительно немного пропусков также в столбце year_of_release (275). Самое большое количество пропусков в столбцах rating (6871), critic_score (8714) и user_score (9268). Данные столбцы необходимо обработать (либо заполнить данные пропуски, либо удалить их), т.к. в случае оставления их в таком виде, полученные результаты могут быть некорректными.

Можно предположить, что поскольку в представленных данных имеются игры, датированные ранее 1980 года, то в то время просто не существовало рейтинга ESRB, а также не было оценок критиков и пользователей.

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

*Пропуски в столбцах genre, name (по 2 пропуска в каждом) логичнее удалить, т.к. эти данные не повлияют на исследования.*

In [17]:
df = df.dropna(subset=['name', 'genre'])

*Пропуски в столбцах eu_sales и jp_sales заменим на среднее значение в зависимости от названия платформы и года выхода игры*

In [18]:
for sales in ['eu_sales', 'jp_sales']:
    group_means = df.groupby(['platform', 'year_of_release'])[sales].transform('mean')
    df[sales] = df[sales].fillna(group_means)

*Пропуски в столбце year_of_release считаю корректным удалить, т.к. это не повлияет на статистику*

In [19]:
df = df.dropna(subset=['year_of_release'])

*Пропуски в столбцах user_score и critic_score заменим на "-1", т.к. маловероятно, что рейтинг может содержать отрицательные значения, при этом значение "0" вполне вероятно*

In [20]:
df['user_score'] = df['user_score'].fillna(-1)

In [21]:
df['critic_score'] = df['critic_score'].fillna(-1)

*Так как столбец rating содержит строковый тип данных, то мы можем заменить пропуски на "нет рейтинга"*

In [22]:
df['rating'] = df['rating'].fillna('нет_рейтинга')

In [23]:
df.isna().sum()

name               0
platform           0
year_of_release    0
genre              0
na_sales           0
eu_sales           0
jp_sales           0
other_sales        0
critic_score       0
user_score         0
rating             0
dtype: int64

По результатам работы с пропусками в данных часть изних была заменена на индикаторы, некоторые пропуски, которые не влияют на последующее изучение данных, были удалены, другая часть была заменена на среднее значение.

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

*Проведем проверку платформ, где выведем все наименования и уникальное количество*

In [ ]:
unique_platform = df['platform'].unique()
display(f'Уникальные значения: {unique_platform}')
unique_platform_count = df['platform'].nunique()
display(f'Количество уникальных значений: {unique_platform_count}')

*Проведем проверку рейтинга игр, где выведем все возможные рейтинги и уникальное количество*

In [ ]:
unique_rating = df['rating'].unique()
display(f'Уникальные значения: {unique_rating}')
unique_rating_count = df['rating'].nunique()
display(f'Количество уникальных значений: {unique_rating_count}')

*Проведем проверку года выпуска игр, где выведем все возможные годы и уникальное количество*

In [ ]:
unique_year_of_release = df['year_of_release'].unique()
display(f'Уникальные значения: {unique_year_of_release}')
unique_year_of_release_count = df['year_of_release'].nunique()
display(f'Количество уникальных значений: {unique_year_of_release_count}')

*Проведем проверку наименований игр, где выведем все возможные годы и уникальное количество*

In [ ]:
unique_name = df['name'].unique()
display(f'Уникальные значения: {unique_name}')
unique_name_count = df['name'].nunique()
display(f'Количество уникальных значений: {unique_name_count}')

*Проведем проверку игровых платформ, где выведем все возможные годы и уникальное количество*

In [ ]:
unique_platform = df['platform'].unique()
display(f'Уникальные значения: {unique_platform}')
unique_platform_count = df['platform'].nunique()
display(f'Количество уникальных значений: {unique_platform_count}')

*Проведем проверку игровых жанров. Для этого выведем все жанры, а также уникальное количество*

In [ ]:
unique_genre = df['genre'].unique()
display(f'Уникальные значения: {unique_genre}')
unique_genre_count = df['genre'].nunique()
display(f'Количество уникальных значений: {unique_genre_count}')

По результатам мы видим, что в наименованиях жанрах существуют неявные дубликаты, которые появились из-за разного написания жанров (разный регистр). Поэтому корректнее будет привести все жанры к одному стилю. Для этого мы сначала сохраним существующие уникальные значения, а после удалим неявные дубликаты.

In [ ]:
unique_genre_all = df['genre'].nunique()

In [ ]:
df['genre'] = df['genre'].str.lower()

In [ ]:
unique_genre = df['genre'].unique()
display(f'Уникальные значения: {unique_genre}')
unique_genre_count = df['genre'].nunique()
display(f'Количество уникальных значений: {unique_genre_count}')

In [ ]:
unique_genre_str = df['genre'].nunique()

In [ ]:
display(df)

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [ ]:
df[df.duplicated()]

Определим количество строк до удаления дубликатов и удалим дубликаты

In [ ]:
str_all = len(df)
display(f'Количество строк до удаления: {str_all}')
cleaned_str = df.drop_duplicates()

In [ ]:
df[df.duplicated(subset=['name', 'address'])]

- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

В процессе работы с данными были обнаружены неявные и явные дубликаты. Ввиду неправильного написания жанров (регистр) были замечены неявные дубликаты в столбце genre (12 шт.), которые были преобразаны в корректный формат. Явные дубликаты были также отсортированы и удалены.

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

*Посчтичаем абсолютное и относительное количество удаленных строк в неявных и явных дубликатах*

In [ ]:
str_all = len(tmp)
display(f'Было строк в исходном датасете: {str_all}')
removed_count = str_all - len(cleaned_str)
str_first = str_all - removed_count
loss_percent = (removed_count / str_all) * 100
display(f'Осталось строк в датасете после обработки: {str_first}')
display(f'Удалено строк в датасете после обработки: {removed_count}')
display(f'Процент потерь: {loss_percent:.2f}')

- После проведения предобработки данных напишите общий промежуточный вывод.

После проведения предобработки данных можно сделать следующие выводы и выполнены следующие работы. Прежде всего, все столбцы были приведены к единообразию (для более удобного чтения). Были определены типы данных в представленных для исследования данных и проведена работа с ними, а именно: в типы данных, которые были некорректные, внесены правки, они приведены к соответствующему типу. Далее были обработаны пропуски и дубликаты. Часть пропусков заменена на индикаторы, некоторые удалены, а некоторые заменены на среднее. Дубликаты были также обработы соответствующим образом: неявные дубликаты были приведены в соответствие, а явные дубликаты были удалены.

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [ ]:
df_actual = cleaned_str[(cleaned_str['year_of_release'] >= 2000) & (cleaned_str['year_of_release'] <= 2013)]

In [ ]:
# Комментарий ревьюера
df.duplicated().sum()

In [ ]:
df_actual.head()

Проведена фильтрация, где были отобраны данные за период с 2000 по 2013 год включительно.

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [ ]:
df_actual = df_actual.copy()
df_actual['user_rating'] = pd.cut(
    df_actual['user_score'],
    bins=[-1, 0, 3, 8, 10],
    labels=['нет оценки','низкая оценка', 'средняя оценка', 'высокая оценка'],
    include_lowest=True,
    right=False
)
result_user = df_actual.groupby('user_rating')['user_score'].count()
display(result_user)

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [ ]:
df_actual = df_actual.copy()
df_actual['critic_rating'] = pd.cut(
    df_actual['critic_score'],
    bins=[-1, 0, 30, 80, 100],
    labels=['нет оценки', 'низкая оценка', 'средняя оценка', 'высокая оценка'],
    include_lowest=True,
    right=False
)
result_critic = df_actual.groupby('critic_rating')['critic_score'].count()
display(result_critic)

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [ ]:
top_7_platforms = df_actual.groupby('platform', as_index=False)['name'].count()
top_7_platforms = top_7_platforms.sort_values(by='name', ascending=False).head(7)
display(top_7_platforms.style.hide_index())

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

В ходе предобработки данных был проанализирован датасет, который состоял из 16 956 строк и 11 столбцов.
При первичном знакомстве с датасетом были обнаружены пропуски в следующих столбцах:
- Critic Score (8242)
- User Score (10152)
- Rating (10085) 
- Year of Release (275)
- Name и Genre (2)
Также было отмечено, что ряд столбцов, а именно Year of Release, EU sales, JP sales и User Score имеют недочеты и отражены в датасете не тем типом данных. Числовые столбцы (продажи и оценки) и год выпуска были приведены к типу float64.

Для более корректной последующей работы все столлбцы были приведены к snak case и все стали соответствовать общепринятому стандарту Python.
При работе с пропусками и дубликатами были выполнены следующие действия и получены следующие результаты:
1. Наибольшее число пропусков зафиксировано в столбцах с оценками критиков и пользователей, рейтингом.
Данные пропуски были обработы следующим образом:
- пропуски в столбцах genre, name и year_of_release были удалены.
- пропуски в столбцах eu_sales и jp_sales были заменены на среднее значение в зависимости от названия платформы и года выхода игры
- пропуски в столбцах user_score и critic_score были заменены на индикаторы (-1)
- пропуски в столбце rating были заменены на "нет рейтинга".
2. Явные и неявные дубликаты были обработаны следующим образом:
- в столбце genre неявные дубликаты существовали из-за разного регистра.  Столбец был приведен к одному стилю и мы увидели, что из 24 существующих жанров осталось лишь 12 уникальных значений.
- 235 явных дубликатов были удалены.
Также была проведена фильтрация данных по году выпуска игр, а именно были отобраны значения, подпадающие во временной промежуток с 2000 по 2013 гг.
Дополнительно была проведена категоризация данных, а именно:
- для столбцов с оценками пользователей и критиков (user_score и critic_score) были созданы рейтинги, включающие в себя категории с "высокая оценка", "средняя оценка", "низкая оценка" и "нет оценки".
- было отобрано топ-7 платформ по количеству игр, выпущенных за весь актуальный период
По итогу данные были приведены к надлежащему виду, что позволяет изучить игровую индустрию и дает возможность корректно подготовить статью о развитии индустрии игр в начале XXI века. 